### PR TITLE
fix(api): add ibis.map and .struct

### LINKED ISSUE #3118: feat: allow dicts in ibis.literal for structs
Currently a struct must be an `OrderedDict`, but since 3.6 there's no reason we can't allow at least `dict`s any possibly any mapping:

```
>>> import ibis
>>> ibis.literal({"a": "b", "c": "d"}, type="struct<a: string, c: string>")
TypeError: Value {'a': 'b', 'c': 'd'} cannot be safely coerced to struct<a: string, c: string>
```

### PR DESCRIPTION
**Pull request by saulpw**
Closes #3118.

### CODE REVIEW

File: ibis/expr/datatypes.py [hunk 1]
```diff
@@ -1414,6 +1414,24 @@ def can_cast_string_to_temporal(
...
+    )
+
+
+@castable.register(Struct, Struct)
+def can_cast_struct(source, target, **kwargs):
+    source_pairs = source.pairs
+    target_pairs = target.pairs
+    for k in {*source.names, *target.names}:
+        if k in target_pairs:
+            if not castable(source_pairs[k], target_pairs[k]):
```
Comment (cpcloud) [inline ibis/expr/datatypes.py:1430]:
```suggestion
    for name in {*source.names, *target.names}:
        if name in target_pairs:
            if not castable(source_pairs[name], target_pairs[name]
```

File: ibis/expr/types.py [hunk 2]
```diff
@@ -1247,7 +1248,7 @@ def literal(value, type=None):
         return ops.Literal(value, dtype=dtype).to_expr()
 
 
-def array(values):
+def array(values, type=None):
```
Comment (cpcloud) [inline ibis/expr/types.py:1251]: Can you add a test case for `array(..., type="some_type_string")`?

**Review by cpcloud — CHANGES_REQUESTED** [review]
Small bit about `k` -> `name`. Thanks, looks great!

File: ibis/expr/types.py (hunk 2 above)
Comment (saulpw) [reply ibis/expr/types.py:1251]: done!

**Review by cpcloud — APPROVED** [review]
Sweet. This looks great.

**Pull request merged**